File: LeetCode_Solution/8. String to Integer (atoi).java
class Solution {
    public int myAtoi(String str) {
        
        // 1. skip heading whitespaces
        str = str.trim();

        int i = 0;
        int n = str.length();
        
        // 2. handle sign
        int sign = 1;
        if (i < n && str.charAt(i) == '+') {
            i++;
            if (str.length() - 1 < i || !Character.isDigit(str.charAt(i))) {        
                return 0;
            }
        }
        if (i < n && str.charAt(i) == '-') {
            sign = -1;
            i++;
            if (str.length() - 1 < i || !Character.isDigit(str.charAt(i))) {        
                return 0;
            }
        }
        
        // 3. get number by digit
        long num = 0;
        while (i < n && Character.isDigit(str.charAt(i))) {
            num = num * 10 + Character.getNumericValue(str.charAt(i));
            i++;
            
            // handle overflow [-2147483648, 2147483647]
            if (num > Integer.MAX_VALUE) {
                return sign == 1 ? Integer.MAX_VALUE : Integer.MIN_VALUE;
            }
            
        }
        
        int result = (int)num;
        return sign * result;
        
    }
}

// time : O(n)
// space: O(1)

File: python_solution/8_String_to_Integer_(atoi).py
class Solution(object):
    def myAtoi(self, s):
        """
        :type s: str
        :rtype: int
        """
        result = 0
        sign = 1
        i = 0
        n = len(s)

        # skip heading spaces
        while i < n and s[i] == ' ':
            i += 1

        # confirm sign
        if i < n and s[i] == '+':
            i += 1
        elif i < n and s[i] == '-':
            sign = -1
            i += 1
        
        # tranverse digits
        while i < n and s[i].isdigit():
            print(s[i])
            digit = int(s[i])
            if self.isOverflow(result, digit):
                return pow(2, 31) - 1 if sign > 0 else -pow(2, 31)
            result = result * 10 + digit
            i += 1
        return sign * result

    def isOverflow(self, n, digit):
        int_max = pow(2, 31) - 1
        if ((n > int_max // 10) or (n == int_max // 10 and digit > int_max % 10)):
            return True
        else:
            return False

